### ISSUE TITLE
The serializer registered with RegisterOverride is not used

### ISSUE THREAD

**Issue by housei** [msgpack/msgpack-cli#159]
Hi.

RegisterOverride was serializer is not used .
It will not be used Test_Test02Serializer.UnpackFromCore specifically .
Test_Test01Serializer and Test_Test02Serializer was created by mpu.
If you change the order of RegisterOverride it seems to work fine .

Is this a bug ?

Unity5.3.4f1
Mac OSX 10.10.5
MsgPack.Cli.0.6.8

https://github.com/housei/MsgPackTest

```csharp
using UnityEngine;
using System.Collections;
using MsgPack;
using MsgPack.Serialization;

namespace Test
{
    public class Test01
    {
        public string id;
        public Test02 test02;
    }

    public class Test02
    {
        public string id;
    }

    public class TestMain : MonoBehaviour 
    {
    	void Start () {
            SerializationContext context = 
                new SerializationContext();

            // RegisterOverride Serializer
            var test_test01serializer = new Test_Test01Serializer(context);
            context.Serializers.RegisterOverride(test_test01serializer);

            var test_test02serializer = new Test_Test02Serializer(context);
            context.Serializers.RegisterOverride(test_test02serializer);

            context.CompatibilityOptions.PackerCompatibilityOptions = PackerCompatibilityOptions.None;

            MessagePackSerializer.Get<Test01>(context);

            Test01 test01 = new Test01(){ 
                id = "test01", 
                test02 = new Test02(){ id = "test02" }
            };

            byte[] bytes =
                MessagePackSerializer.Get<Test01>(context).PackSingleObject(test01);

            Test01 __test01 =
                MessagePackSerializer.Get<Test01>(context).UnpackSingleObject(bytes);
            
            Debug.Log(__test01.id + " " + __test01.test02.id);
    	}
    }
}
```

**Comment by yfakariya** [comment]
Thank you for feedback, and I edited the title to be easy to find this problem from others.
I'll try to test your code.

**Comment by yfakariya** [comment]
Sorry for delay. Did you say that:
* You have two types and one owns another.
* You saw that if you call `RegisterOverride` for outer type (`Test01`) then inner type (`Test02`), the `Test_Test02Serializer` was not used. However, if you call `RegisterOverride` for  inner type (`Test02`) then outer type (`Test01`), the `Test_Test02Serializer` was used.
If so, this is known issue and is by design, but there is a solution. 

Problem details:
Generated serializers cache dependent serializers on their fields for performance, so `Test01` constructor calls `SerializationContext.GetSerializer<Test02>()` then the serializer engine creates "Reflection-based" serializer and caches it in its field. 

Solution:
This problem was resolved in #103. You can use `SerializationContext.ResolveSerializer` event for this issue. Specifically, in the event handler, you should do as following:
1. Check `ResolveSerializerEventArgs.TargetType` property which contains "resolving" type for serializer.
1. Instantiate appropriate serializer instance for `TargetType`.
1. Set the instance via `ResolveSerializerEventArgs.SetSerializer`.

**Comment by housei** [comment]
I was able to solve !
Thank you.
